File: scripts/playwright_smoke.js
export default async function runMeddraSmoke(page) {
  const checks = [];
  const consoleErrors = [];
  page.on("console", (message) => {
    if (message.type() === "error") consoleErrors.push(message.text());
  });

  const ensureVisible = async (name, locator) => {
    await locator.first().waitFor({ state: "visible", timeout: 15000 });
    checks.push(name);
  };

  const ensureNoResultTextOverlap = async (name) => {
    const issues = await page.locator(".center-pane .result").evaluateAll((rows) => {
      const selectors = [
        [".result-main .level-badge", "层级标记"],
        [".result-main strong", "主名称"],
        [".result-main small", "英文名"],
        [".result-main em", "代码"],
        [".result-meta span:nth-child(1)", "匹配类型"],
        [".result-meta span:nth-child(2)", "匹配字段"],
        [".result-meta p", "匹配说明"],
        [".result-actions", "操作区"],
      ];
      const visibleRect = (element) => {
        const rect = element.getBoundingClientRect();
        const style = window.getComputedStyle(element);
        if (style.display === "none" || style.visibility === "hidden" || rect.width <= 1 || rect.height <= 1) {
          return null;
        }
        return {
          left: rect.left,
          right: rect.right,
          top: rect.top,
          bottom: rect.bottom,
          width: rect.width,
          height: rect.height,
        };
      };
      const overlaps = [];
      rows.forEach((row, rowIndex) => {
        if (row.scrollWidth > row.clientWidth + 4) {
          overlaps.push({
            row: rowIndex + 1,
            type: "row-overflow",
            width: row.clientWidth,
            scrollWidth: row.scrollWidth,
          });
        }
        const parts = selectors.flatMap(([selector, label]) => {
          return Array.from(row.querySelectorAll(selector)).map((element, partIndex) => {
            const rect = visibleRect(element);
            if (!rect) return null;
            return {
              label: partIndex ? `${label}#${partIndex + 1}` : label,
              text: (element.textContent || "").trim().replace(/\s+/g, " ").slice(0, 32),
              rect,
            };
          }).filter(Boolean);
        });
        for (let i = 0; i < parts.length; i += 1) {
          for (let j = i + 1; j < parts.length; j += 1) {
            const a = parts[i];
            const b = parts[j];
            const overlapX = Math.min(a.rect.right, b.rect.right) - Math.max(a.rect.left, b.rect.left);
            const overlapY = Math.min(a.rect.bottom, b.rect.bottom) - Math.max(a.rect.top, b.rect.top);
            if (overlapX > 2 && overlapY > 2) {
              overlaps.push({
                row: rowIndex + 1,
                type: "text-overlap",
                a: `${a.label}:${a.text}`,
                b: `${b.label}:${b.text}`,
                overlapX: Math.round(overlapX),
                overlapY: Math.round(overlapY),
              });
            }
          }
        }
      });
      return overlaps.slice(0, 6);
    });
    if (issues.length) {
      throw new Error(`${name}存在文字碰撞或横向溢出: ${JSON.stringify(issues)}`);
    }
    checks.push(name);
  };

  const appUrl = typeof process !== "undefined" && process.env.MEDDRA_BROWSER_URL
    ? process.env.MEDDRA_BROWSER_URL
    : "http://127.0.0.1:8765/";
  await page.goto(appUrl, { waitUntil: "networkidle" });
  await page.evaluate(() => {
    localStorage.removeItem("meddra.bin");
    localStorage.removeItem("meddra.history");
    localStorage.removeItem("meddra.panes");
  });
  await page.reload({ waitUntil: "networkidle" });
  await ensureVisible("中文主导航", page.locator(".module-nav button.active", { hasText: "搜索" }));
  await ensureVisible("品牌Logo", page.locator(".brand img"));
  const logoLoaded = await page.locator(".brand img").evaluate((node) => node instanceof HTMLImageElement && node.complete && node.naturalWidth >= 64);
  if (!logoLoaded) {
    throw new Error("品牌Logo未正确加载");
  }
  checks.push("品牌Logo加载成功");
  await ensureVisible("SOC层级树", page.getByText("SOC层级"));
  await ensureVisible("版本选择", page.locator(".version-select"));
  await ensureVisible("默认PT搜索层级", page.locator(".level-filter button.active", { hasText: "PT" }));
  await ensureVisible("SMQ搜索层级可选", page.locator(".level-filter button", { hasText: "SMQ" }));
  if (await page.getByRole("button", { name: "代码查询" }).count()) {
    throw new Error("旧代码查询导航仍然存在");
  }
  const versionOptions = await page.locator(".version-select option").allTextContents();
  if (!versionOptions.some((value) => value.includes("MedDRA 29.0"))) {
    throw new Error("版本下拉框未包含MedDRA 29.0测试夹具");
  }
  if (!versionOptions.every((value) => value.includes("双语") || value.includes("仅中文") || value.includes("仅英文"))) {
    throw new Error("版本下拉框未显示语言可用性标签");
  }
  await page.screenshot({ path: "output/playwright/desktop-initial.png", fullPage: true });

  await page.getByRole("button", { name: "英文" }).click();
  await ensureVisible("英文显示模式", page.locator(".segmented button.active", { hasText: "英文" }));
  await page.getByRole("button", { name: "双语" }).click();

  const searchInput = page.getByPlaceholder("可输入AE/MH名称进行模糊查询或输入代码进行精确查询");
  await searchInput.fill("rhabdomyolisys");
  await page.locator(".center-pane .query-bar button.primary").click();
  await ensureVisible("模糊查询分组", page.getByText("模糊候选").first());
  await ensureVisible("模糊候选命中Rhabdomyolysis", page.getByText("Rhabdomyolysis").first());
  await ensureVisible("模糊原因无分数", page.getByText(/近似文本候选/).first());
  const fuzzyMeta = await page.locator(".result.fuzzy .result-meta").first().textContent();
  if (/模糊相似度|\b\d{2,3}(?:\.\d)?\b/.test(fuzzyMeta || "")) {
    throw new Error(`模糊结果仍显示分数: ${fuzzyMeta}`);
  }
  await ensureNoResultTextOverlap("模糊搜索结果无文字碰撞");

  await page.locator(".result-main").first().click();
  await ensureVisible("中心详情关系模块", page.locator(".center-pane .detail-workspace"));
  await ensureVisible("右侧关系树", page.locator(".right-pane .relationship-tree-panel"));
  await ensureVisible("关系树当前节点高亮", page.locator(".right-pane .relationship-tree-node.current", { hasText: "Rhabdomyolysis" }));
  await ensureVisible("关系树父级SOC展开", page.locator(".right-pane .relationship-tree-node", { hasText: "Musculoskeletal" }));
  await ensureVisible("关系树直接子级默认折叠", page.locator(".right-pane .relationship-tree-toggle", { hasText: "直接子级" }));
  await ensureVisible("父系路径默认展开", page.getByText("父系层级路径"));
  await ensureVisible("术语详情SMQ关系", page.getByText("SMQ成员关系"));
  await page.locator(".right-pane .relationship-tree-toggle", { hasText: "直接子级" }).first().click();
  await ensureVisible("关系树子级可展开", page.locator(".right-pane .relationship-tree-children .relationship-tree-node", { hasText: "LLT" }).first());
  await page.screenshot({ path: "output/playwright/v3-detail-tree.png", fullPage: true });
  await page.locator(".module-nav button", { hasText: /^搜索$/ }).click();
  await page.locator('.result-actions button[title="加入Research Bin"]').first().click();
  await ensureVisible("Research Bin已加入状态", page.getByText("已加入").first());

  await page.locator(".module-nav button", { hasText: /^搜索$/ }).click();
  await searchInput.fill("10039020");
  await page.locator(".center-pane .query-bar button.primary").click();
  await ensureVisible("主搜索框代码查询分组", page.locator(".result-group", { hasText: "代码匹配" }));
  await ensureVisible("主搜索框代码查询结果", page.getByText("横纹肌溶解").first());
  await ensureNoResultTextOverlap("代码搜索结果无文字碰撞");

  const leftBefore = await page.locator(".left-pane").boundingBox();
  const leftResizer = await page.locator(".pane-resizer-left").boundingBox();
  if (!leftBefore || !leftResizer) throw new Error("未找到左侧拖拽栏");
  await page.mouse.move(leftResizer.x + leftResizer.width / 2, leftResizer.y + 60);
  await page.mouse.down();
  await page.mouse.move(leftResizer.x + 84, leftResizer.y + 60, { steps: 8 });
  await page.mouse.up();
  const leftAfter = await page.locator(".left-pane").boundingBox();
  if (!leftAfter || leftAfter.width <= leftBefore.width + 36) {
    throw new Error(`左侧栏拖拽未改变宽度: before=${leftBefore?.width} after=${leftAfter?.width}`);
  }
  checks.push("三栏左侧宽度可拖拽调整");
  const rightBefore = await page.locator(".right-pane").boundingBox();
  const rightResizer = await page.locator(".pane-resizer-right").boundingBox();
  if (!rightBefore || !rightResizer) throw new Error("未找到右侧拖拽栏");
  await page.mouse.move(rightResizer.x + rightResizer.width / 2, rightResizer.y + 60);
  await page.mouse.down();
  await page.mouse.move(rightResizer.x - 84, rightResizer.y + 60, { steps: 8 });
  await page.mouse.up();
  const rightAfter = await page.locator(".right-pane").boundingBox();
  if (!rightAfter || rightAfter.width <= rightBefore.width + 36) {
    throw new Error(`右侧栏拖拽未改变宽度: before=${rightBefore?.width} after=${rightAfter?.width}`);
  }
  checks.push("三栏右侧宽度可拖拽调整");
  const primaryButtonBox = await page.locator(".center-pane .query-bar button.primary").boundingBox();
  if (!primaryButtonBox || primaryButtonBox.height > 52) {
    throw new Error(`窄中栏搜索按钮疑似被挤成竖排: height=${primaryButtonBox?.height}`);
  }
  const navHasHorizontalOverflow = await page.locator(".module-nav").evaluate((node) => node.scrollWidth > node.clientWidth + 2);
  if (navHasHorizontalOverflow) {
    throw new Error("窄中栏模块导航存在横向溢出或按钮被截断");
  }
  checks.push("窄中栏按钮换行不竖排");
  await searchInput.fill("横纹肌");
  await page.locator(".center-pane .query-bar button.primary").click();
  await ensureVisible("窄中栏包含匹配结果", page.locator(".result-group", { hasText: "包含匹配" }));
  await ensureNoResultTextOverlap("窄中栏搜索结果无文字碰撞");
  await page.screenshot({ path: "output/playwright/final-narrow-search-results.png", fullPage: true });

  await page.getByRole("button", { name: "高级搜索" }).click();
  const advancedInputs = page.locator('input[placeholder="可输入AE/MH名称、中文/英文片段或代码"]');
  await advancedInputs.nth(0).fill("renal");
  await advancedInputs.nth(1).fill("failure");
  await page.locator(".advanced-grid button.primary").click();
  await ensureVisible("高级搜索结果", page.locator(".result-list .result").first());
  await ensureNoResultTextOverlap("高级搜索结果无文字碰撞");

  await page.locator(".module-nav button", { hasText: /^搜索$/ }).click();
  await page.locator(".level-filter button", { hasText: "SMQ" }).click();
  await searchInput.fill("横纹肌");
  await page.locator(".center-pane .query-bar button.primary").click();
  await ensureVisible("统一搜索SMQ结果", page.locator(".result-group", { hasText: "SMQ匹配" }));
  await ensureNoResultTextOverlap("SMQ搜索结果无文字碰撞");
  await page.locator(".result-group", { hasText: "SMQ匹配" }).locator(".result-main").first().click();
  await ensureVisible("SMQ详情中心展示", page.locator(".center-pane .detail-workspace", { hasText: "SMQ" }));
  await ensureVisible("SMQ关系树当前节点", page.locator(".right-pane .relationship-tree-node.current", { hasText: "横纹肌溶解/肌病" }));
  await ensureVisible("SMQ内容默认折叠", page.locator(".right-pane .relationship-tree-toggle", { hasText: "包含术语" }));
  await page.locator(".right-pane .relationship-tree-toggle", { hasText: "包含术语" }).first().click();
  await ensureVisible("SMQ广义狭义显示", page.getByText(/狭义|广义/).first());
  await page.screenshot({ path: "output/playwright/v3-smq-tree.png", fullPage: true });
  const smqDownload = await Promise.all([
    page.waitForEvent("download"),
    page.getByRole("button", { name: /导出SMQ/ }).click(),
  ]);
  await smqDownload[0].saveAs("output/playwright/smq-export.csv");

  await page.locator(".module-nav button", { hasText: "Research Bin" }).click();
  await ensureVisible("Research Bin列表", page.locator(".result-list .result").first());
  await ensureVisible("Research Bin移除按钮", page.getByRole("button", { name: /移除/ }).first());
  await ensureNoResultTextOverlap("Research Bin结果无文字碰撞");
  const binDownload = await Promise.all([
    page.waitForEvent("download"),
    page.getByRole("button", { name: /导出JSON/ }).click(),
  ]);
  await binDownload[0].saveAs("output/playwright/research-bin.json");

  await page.getByRole("button", { name: "设置" }).click();
  await ensureVisible("词典来源导入", page.getByText("词典来源导入"));
  await ensureVisible("导入文件提示", page.getByText("mdhier.asc"));
  const synonymResponse = page.waitForResponse((response) => response.url().includes("/api/synonyms") && response.ok());
  await page.getByRole("button", { name: "查看中文同义词表" }).click();
  await synonymResponse;
  await ensureVisible("同义词表状态", page.locator(".synonym-status"));

  await page.setViewportSize({ width: 390, height: 820 });
  await page.screenshot({ path: "output/playwright/desktop-narrow-settings.png", fullPage: true });
  const bodyWidth = await page.evaluate(() => document.documentElement.scrollWidth);
  const viewportWidth = await page.evaluate(() => window.innerWidth);
  if (bodyWidth > viewportWidth + 4) {
    throw new Error(`窄屏存在页面级横向溢出: body=${bodyWidth}, viewport=${viewportWidth}`);
  }
  checks.push("桌面窄窗口无页面级横向溢出");

  if (consoleErrors.length) {
    throw new Error(`浏览器console错误: ${consoleErrors.join(" | ")}`);
  }
  return { checks, downloads: ["output/playwright/smq-export.csv", "output/playwright/research-bin.json"] };
}

File: scripts/playwright_smoke.py
from __future__ import annotations

import os
import re
from pathlib import Path

from playwright.sync_api import Browser, Page, expect, sync_playwright


ROOT = Path(__file__).resolve().parents[1]
OUTPUT = ROOT / "output" / "playwright"
APP_URL = os.environ.get("MEDDRA_BROWSER_URL", "http://127.0.0.1:8765/")
SEARCH_PLACEHOLDER = "可输入AE/MH名称进行模糊查询或输入代码进行精确查询"
ADVANCED_PLACEHOLDER = "可输入AE/MH名称、中文/英文片段或代码"


def check_result_layout(page: Page, label: str) -> None:
    issues = page.locator(".center-pane .result").evaluate_all(
        """(rows) => {
          const selectors = [
            [".result-main .level-badge", "level"],
            [".result-main strong", "name"],
            [".result-main small", "subtitle"],
            [".result-main em", "code"],
            [".result-meta span:nth-child(1)", "match"],
            [".result-meta span:nth-child(2)", "field"],
            [".result-meta p", "reason"],
            [".result-actions", "actions"]
          ];
          const visibleRect = (element) => {
            const rect = element.getBoundingClientRect();
            const style = window.getComputedStyle(element);
            if (style.display === "none" || style.visibility === "hidden" || rect.width <= 1 || rect.height <= 1) return null;
            return { left: rect.left, right: rect.right, top: rect.top, bottom: rect.bottom, width: rect.width, height: rect.height };
          };
          const overlaps = [];
          rows.forEach((row, rowIndex) => {
            if (row.scrollWidth > row.clientWidth + 4) {
              overlaps.push({ row: rowIndex + 1, type: "row-overflow", width: row.clientWidth, scrollWidth: row.scrollWidth });
            }
            const parts = selectors.flatMap(([selector, name]) => Array.from(row.querySelectorAll(selector)).map((element, index) => {
              const rect = visibleRect(element);
              if (!rect) return null;
              return { label: index ? `${name}#${index + 1}` : name, text: (element.textContent || "").trim().slice(0, 32), rect };
            }).filter(Boolean));
            for (let i = 0; i < parts.length; i += 1) {
              for (let j = i + 1; j < parts.length; j += 1) {
                const a = parts[i];
                const b = parts[j];
                const overlapX = Math.min(a.rect.right, b.rect.right) - Math.max(a.rect.left, b.rect.left);
                const overlapY = Math.min(a.rect.bottom, b.rect.bottom) - Math.max(a.rect.top, b.rect.top);
                if (overlapX > 2 && overlapY > 2) {
                  overlaps.push({ row: rowIndex + 1, type: "text-overlap", a: `${a.label}:${a.text}`, b: `${b.label}:${b.text}` });
                }
              }
            }
          });
          return overlaps.slice(0, 8);
        }"""
    )
    if issues:
        raise AssertionError(f"{label}: text overlap or horizontal overflow: {issues}")


def expected_center_min(viewport_width: int) -> int:
    if viewport_width <= 1100:
        return 0
    if viewport_width >= 1700:
        return 900
    if viewport_width >= 1300:
        return round(viewport_width * 0.52)
    return round(viewport_width * 0.5)


def inspect_result_density(page: Page, label: str, viewport_width: int) -> None:
    metrics = page.evaluate(
        """() => {
          const center = document.querySelector(".center-pane");
          const group = document.querySelector(".center-pane .result-group");
          const row = document.querySelector(".center-pane .result");
          const main = document.querySelector(".center-pane .result-main");
          if (!center || !group || !row || !main) return { missing: true };
          const centerRect = center.getBoundingClientRect();
          const groupRect = group.getBoundingClientRect();
          const rowRect = row.getBoundingClientRect();
          const mainRect = main.getBoundingClientRect();
          const style = window.getComputedStyle(row);
          const columns = style.gridTemplateColumns.split(" ").filter(Boolean).length;
          return {
            missing: false,
            viewport: window.innerWidth,
            bodyWidth: document.documentElement.scrollWidth,
            centerWidth: centerRect.width,
            groupWidth: groupRect.width,
            rowHeight: rowRect.height,
            rowWidth: rowRect.width,
            rowScrollWidth: row.scrollWidth,
            mainWidth: mainRect.width,
            columns
          };
        }"""
    )
    if metrics.get("missing"):
        raise AssertionError(f"{label}: result metrics missing")
    if metrics["bodyWidth"] > metrics["viewport"] + 4:
        raise AssertionError(f"{label}: page overflow {metrics}")
    if metrics["rowScrollWidth"] > metrics["rowWidth"] + 4:
        raise AssertionError(f"{label}: result row overflow {metrics}")
    if metrics["groupWidth"] < metrics["centerWidth"] - 36:
        raise AssertionError(f"{label}: result group does not use center width {metrics}")
    min_center = expected_center_min(viewport_width)
    if min_center and metrics["centerWidth"] < min_center - 28:
        raise AssertionError(f"{label}: center pane below responsive budget {metrics}, expected >= {min_center}")
    if metrics["centerWidth"] >= 660:
        if metrics["columns"] < 3:
            raise AssertionError(f"{label}: wide result row collapsed into stacked layout {metrics}")
        if metrics["rowHeight"] > 118:
            raise AssertionError(f"{label}: wide result row too tall/crumpled {metrics}")
        if metrics["mainWidth"] < min(360, metrics["centerWidth"] * 0.48):
            raise AssertionError(f"{label}: main result area too narrow {metrics}")


def drag_pane(page: Page, side: str, delta: int) -> None:
    locator = page.locator(f".pane-resizer-{side}")
    box = locator.bounding_box()
    if not box:
        return
    x = box["x"] + box["width"] / 2
    y = box["y"] + 120
    page.mouse.move(x, y)
    page.mouse.down()
    page.mouse.move(x + delta, y, steps=10)
    page.mouse.up()
    page.wait_for_timeout(120)


def ensure_search_results(page: Page) -> None:
    page.get_by_role("navigation").get_by_role("button", name="搜索", exact=True).click()
    search = page.get_by_placeholder(SEARCH_PLACEHOLDER)
    search.fill("横纹肌")
    page.locator(".center-pane .query-bar button.primary").click()
    expect(page.locator(".result-group", has_text="包含匹配")).to_be_visible(timeout=15000)


def check_responsive_matrix(page: Page) -> list[str]:
    scenarios: list[str] = []
    for width in [980, 1180, 1366, 1440, 1600, 1920]:
        page.set_viewport_size({"width": width, "height": 900})
        page.evaluate(
            """() => {
              localStorage.removeItem("meddra.panes");
              localStorage.removeItem("meddra.history");
            }"""
        )
        page.reload(wait_until="networkidle")
        expect(page.get_by_role("navigation").get_by_role("button", name="搜索", exact=True)).to_have_class(
            re.compile(r"\bactive\b"), timeout=15000
        )
        ensure_search_results(page)
        check_result_layout(page, f"responsive {width} default")
        inspect_result_density(page, f"responsive {width} default", width)
        if width > 1100:
            for label, actions in [
                ("left-wide", [("left", 260)]),
                ("right-wide", [("right", -260)]),
                ("both-attempted-wide", [("left", 260), ("right", -260)]),
                ("both-narrowed", [("left", -180), ("right", 180)])
            ]:
                for side, delta in actions:
                    drag_pane(page, side, delta)
                ensure_search_results(page)
                check_result_layout(page, f"responsive {width} {label}")
                inspect_result_density(page, f"responsive {width} {label}", width)
        page.screenshot(path=str(OUTPUT / f"responsive-{width}.png"), full_page=True)
        scenarios.append(f"responsive matrix {width}px")
    return scenarios


def run(page: Page) -> list[str]:
    checks: list[str] = []
    console_errors: list[str] = []
    page.on("console", lambda message: console_errors.append(message.text) if message.type == "error" else None)

    page.goto(APP_URL, wait_until="networkidle")
    page.evaluate(
        """() => {
          localStorage.removeItem("meddra.bin");
          localStorage.removeItem("meddra.history");
          localStorage.removeItem("meddra.panes");
        }"""
    )
    page.reload(wait_until="networkidle")

    expect(page.locator(".module-nav button.active", has_text="搜索")).to_be_visible(timeout=15000)
    expect(page.get_by_text("建议最大化窗口后使用")).to_be_visible()
    expect(page.locator(".brand img")).to_be_visible()
    logo_loaded = page.locator(".brand img").evaluate("(node) => node.complete && node.naturalWidth >= 64")
    assert logo_loaded, "brand logo did not load"
    checks.append("logo and maximized-window tip loaded")

    expect(page.locator(".version-select")).to_be_visible()
    options = page.locator(".version-select option").all_text_contents()
    assert any("MedDRA 29.0" in item for item in options), options
    assert all(("双语" in item or "仅中文" in item or "仅英文" in item) for item in options), options
    checks.append("version selector includes language labels")

    expect(page.locator(".level-filter button.active", has_text="PT")).to_be_visible()
    expect(page.locator(".level-filter button", has_text="SMQ")).to_be_visible()
    OUTPUT.mkdir(parents=True, exist_ok=True)
    page.screenshot(path=str(OUTPUT / "logo-desktop.png"), full_page=True)

    page.get_by_role("button", name="英文").click()
    expect(page.locator(".segmented button.active", has_text="英文")).to_be_visible()
    page.get_by_role("button", name="双语").click()
    checks.append("database display mode switches")

    search = page.get_by_placeholder(SEARCH_PLACEHOLDER)
    expect(search).to_have_value("")
    checks.append("search placeholder without example value")
    search.fill("rhabdomyolisys")
    page.locator(".center-pane .query-bar button.primary").click()
    expect(page.get_by_text("模糊候选").first).to_be_visible(timeout=15000)
    expect(page.get_by_text("Rhabdomyolysis").first).to_be_visible()
    fuzzy_meta = page.locator(".result.fuzzy .result-meta").first.text_content() or ""
    assert "模糊相似度" not in fuzzy_meta, fuzzy_meta
    check_result_layout(page, "fuzzy search")
    checks.append("fuzzy search without visible score")

    page.locator(".result-main").first.click()
    expect(page.locator(".center-pane .detail-workspace")).to_be_visible()
    expect(page.locator(".right-pane .relationship-tree-node.current", has_text="Rhabdomyolysis")).to_be_visible()
    expect(page.locator(".right-pane .relationship-tree-node", has_text="Musculoskeletal")).to_be_visible()
    pseudo_line = page.locator(".right-pane .relationship-tree-node.current").first.evaluate(
        """(node) => getComputedStyle(node, "::before").content"""
    )
    assert pseudo_line in ("none", "normal", ""), pseudo_line
    expect(page.locator(".right-pane .relationship-tree-toggle", has_text="直接子级")).to_be_visible()
    page.locator(".right-pane .relationship-tree-toggle", has_text="直接子级").first.click()
    expect(page.locator(".right-pane .relationship-tree-children .relationship-tree-node", has_text="LLT").first).to_be_visible()
    page.screenshot(path=str(OUTPUT / "v3-detail-tree.png"), full_page=True)
    checks.append("detail and relationship tree without connector lines")

    page.get_by_role("button", name="搜索", exact=True).click()
    page.locator('.result-actions button[title="加入Research Bin"]').first.click()
    expect(page.get_by_text("已加入").first).to_be_visible()
    search.fill("10039020")
    page.locator(".center-pane .query-bar button.primary").click()
    expect(page.locator(".result-group", has_text="代码匹配")).to_be_visible()
    expect(page.get_by_text("横纹肌溶解").first).to_be_visible()
    check_result_layout(page, "code search")
    checks.append("code search in unified box")

    left_before = page.locator(".left-pane").bounding_box()
    left_resizer = page.locator(".pane-resizer-left").bounding_box()
    assert left_before and left_resizer, "left pane or resizer missing"
    page.mouse.move(left_resizer["x"] + left_resizer["width"] / 2, left_resizer["y"] + 60)
    page.mouse.down()
    page.mouse.move(left_resizer["x"] + 84, left_resizer["y"] + 60, steps=8)
    page.mouse.up()
    left_after = page.locator(".left-pane").bounding_box()
    assert left_after and left_after["width"] > left_before["width"] + 36, (left_before, left_after)
    checks.append("left pane resize")

    right_before = page.locator(".right-pane").bounding_box()
    right_resizer = page.locator(".pane-resizer-right").bounding_box()
    assert right_before and right_resizer, "right pane or resizer missing"
    page.mouse.move(right_resizer["x"] + right_resizer["width"] / 2, right_resizer["y"] + 60)
    page.mouse.down()
    page.mouse.move(right_resizer["x"] - 84, right_resizer["y"] + 60, steps=8)
    page.mouse.up()
    right_after = page.locator(".right-pane").bounding_box()
    assert right_after and right_after["width"] > right_before["width"] + 36, (right_before, right_after)
    checks.append("right pane resize")

    search.fill("横纹肌")
    page.locator(".center-pane .query-bar button.primary").click()
    expect(page.locator(".result-group", has_text="包含匹配")).to_be_visible()
    check_result_layout(page, "narrow center search")
    page.screenshot(path=str(OUTPUT / "logo-narrow-search.png"), full_page=True)
    checks.append("narrow results no overlap")
    checks.extend(check_responsive_matrix(page))

    page.get_by_role("button", name="高级搜索").click()
    advanced_inputs = page.locator(f'input[placeholder="{ADVANCED_PLACEHOLDER}"]')
    expect(advanced_inputs).to_have_count(2)
    expect(advanced_inputs.nth(0)).to_have_value("")
    expect(advanced_inputs.nth(1)).to_have_value("")
    advanced_inputs.nth(0).fill("renal")
    advanced_inputs.nth(1).fill("failure")
    page.locator(".advanced-grid button.primary").click()
    expect(page.locator(".result-list .result").first).to_be_visible()
    check_result_layout(page, "advanced search")
    checks.append("advanced search")

    page.get_by_role("button", name="搜索", exact=True).click()
    page.locator(".level-filter button", has_text="SMQ").click()
    search.fill("横纹肌")
    page.locator(".center-pane .query-bar button.primary").click()
    expect(page.locator(".result-group", has_text="SMQ匹配")).to_be_visible()
    page.locator(".result-group", has_text="SMQ匹配").locator(".result-main").first.click()
    expect(page.locator(".center-pane .detail-workspace", has_text="SMQ")).to_be_visible()
    expect(page.locator(".right-pane .relationship-tree-node.current", has_text="横纹肌溶解/肌病")).to_be_visible()
    page.locator(".right-pane .relationship-tree-toggle", has_text="包含术语").first.click()
    expect(page.get_by_text("狭义").first).to_be_visible()
    with page.expect_download() as smq_download:
        page.locator("button", has_text="导出SMQ").click()
    smq_download.value.save_as(str(OUTPUT / "smq-export.csv"))
    checks.append("SMQ detail and export")

    page.locator(".module-nav button", has_text="Research Bin").click()
    expect(page.locator(".result-list .result").first).to_be_visible()
    expect(page.locator("button", has_text="移除").first).to_be_visible()
    check_result_layout(page, "research bin")
    with page.expect_download() as bin_download:
        page.locator("button", has_text="导出JSON").click()
    bin_download.value.save_as(str(OUTPUT / "research-bin.json"))
    checks.append("Research Bin export")

    page.get_by_role("button", name="设置").click()
    expect(page.get_by_text("词典来源导入")).to_be_visible()
    expect(page.get_by_text("mdhier.asc")).to_be_visible()
    expect(page.get_by_role("button", name="加入来源")).to_be_visible()
    expect(page.get_by_placeholder("也可手动粘贴词典文件夹路径")).to_be_visible()
    settings_text = page.locator(".settings-panel").text_content() or ""
    assert "/Users/" not in settings_text and ".sqlite" not in settings_text, settings_text
    with page.expect_response(lambda response: "/api/synonyms" in response.url and response.ok):
        page.get_by_role("button", name="查看中文同义词表").click()
    expect(page.locator(".synonym-status")).to_contain_text(re.compile(r"已载入|未找到"))
    checks.append("settings and synonym table")

    page.set_viewport_size({"width": 390, "height": 820})
    page.screenshot(path=str(OUTPUT / "desktop-narrow-settings.png"), full_page=True)
    body_width = page.evaluate("document.documentElement.scrollWidth")
    viewport_width = page.evaluate("window.innerWidth")
    assert body_width <= viewport_width + 4, (body_width, viewport_width)
    checks.append("desktop narrow window no page overflow")

    if console_errors:
        raise AssertionError("browser console errors: " + " | ".join(console_errors))
    return checks


def check_mobile_gate(browser: Browser) -> list[str]:
    context = browser.new_context(
        viewport={"width": 390, "height": 820},
        is_mobile=True,
        has_touch=True,
        user_agent=(
            "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) "
            "AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1"
        ),
    )
    page = context.new_page()
    try:
        page.goto(APP_URL, wait_until="networkidle")
        expect(page.get_by_text("建议使用电脑端打开")).to_be_visible(timeout=15000)
        expect(page.get_by_text("Mac 或 Windows")).to_be_visible()
        page.screenshot(path=str(OUTPUT / "mobile-desktop-recommendation.png"), full_page=True)
        body_width = page.evaluate("document.documentElement.scrollWidth")
        viewport_width = page.evaluate("window.innerWidth")
        assert body_width <= viewport_width + 4, (body_width, viewport_width)
        return ["mobile device shows desktop recommendation"]
    finally:
        context.close()


def main() -> None:
    with sync_playwright() as playwright:
        browser = playwright.chromium.launch(headless=True)
        page = browser.new_page(viewport={"width": 1440, "height": 960}, accept_downloads=True)
        try:
            checks = run(page)
            checks.extend(check_mobile_gate(browser))
        finally:
            browser.close()
    print("Playwright smoke passed:")
    for check in checks:
        print(f"- {check}")


if __name__ == "__main__":
    main()
